python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -2390,6 +2399,38 @@ def get_handlers(self) -> Sequence[logging.Handler]:
     def store_event(self, event: "EventLogEntry") -> None:
         self._event_storage.store_event(event)
 
+    def _should_retry_run(self, run: DagsterRun) -> bool:
+        from dagster._core.events import RunFailureReason
+
+        if run.status == DagsterRunStatus.FAILURE and self.run_retries_enabled:
+            max_retries = (
+                int(run.tags[MAX_RETRIES_TAG])
+                if run.tags.get(MAX_RETRIES_TAG) is not None
+                else self.run_retries_max_retries
+            )
+            if max_retries > 0:
+                run_group = self.get_run_group(run.run_id)
+                if run_group is not None:
+                    _, run_group_iter = run_group
+                    if len(list(run_group_iter)) < max_retries:
+                        retry_on_asset_or_op_failure = get_boolean_tag_value(
+                            run.tags.get(RETRY_ON_ASSET_OR_OP_FAILURE_TAG),
+                            default_value=self.retry_on_asset_or_op_failure,
+                        )
+                        if (
+                            run.tags.get(RUN_FAILURE_REASON_TAG)
+                            == RunFailureReason.STEP_FAILURE.value
+                            and not retry_on_asset_or_op_failure
+                        ):
+                            self.report_engine_event(
+                                "Not retrying run since it failed due to an asset or op failure and run retries "
+                                "are configured with retry_on_asset_or_op_failure set to false.",
+                                run,
+                            )
```

Comment:
It's a little odd that this method markets itself as a boolean but has a potential side effect of writing an event (within another store event call) - I think we may still want to defer that to the daemon, as one of the things it might do in addition to actually retrying the event

Reply:
so in the daemon as we process the runs we see:
- if will_retry is True, launch retry 
- if will_retry is False and `run.tags.get(RUN_FAILURE_REASON_TAG) == RunFailureReason.STEP_FAILURE.value and not retry_on_asset_or_op_failure` then report this engine event

?